jetty-websocket/websocket-core/src/test/java/org/eclipse/jetty/websocket/core/proxy/WebSocketProxy.java

```diff
@@ -0,0 +1,635 @@
+import org.eclipse.jetty.websocket.core.OpCode;
+import org.eclipse.jetty.websocket.core.client.WebSocketCoreClient;
+
+class WebSocketProxy
+{
+    enum State
+    {
+        NOT_OPEN,
+        CONNECTING,
+        OPEN,
+        ICLOSED,
+        OCLOSED,
+        CLOSED,
+        FAILED
+    }
+
+    private final Object lock = new Object();
+
+    WebSocketCoreClient client;
+    private URI serverUri;
+
+    public Client2Proxy client2Proxy = new Client2Proxy();
+    public Server2Proxy server2Proxy = new Server2Proxy();
+
+    public WebSocketProxy(WebSocketCoreClient client, URI serverUri)
+    {
+        this.client = client;
+        this.serverUri = serverUri;
+    }
+    class Client2Proxy implements FrameHandler
+    {
+        private CoreSession client;
+        private State state = State.NOT_OPEN;
+
+        private Callback closeCallback;
+        private Throwable error;
+
+        public BlockingQueue<Frame> receivedFrames = new BlockingArrayQueue<>();
+        protected CountDownLatch closed = new CountDownLatch(1);
+
+        public State getState()
+        {
+            synchronized (this)
+            {
+                return state;
+            }
+        }
+
+        @Override
+        public void onOpen(CoreSession session, Callback callback)
+        {
+            System.err.println(toString() + " onOpen(): " + session);
+
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case NOT_OPEN:
+                        state = State.CONNECTING;
+                        client = session;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            if (failure != null)
+                callback.failed(failure);
+            else
+                server2Proxy.connect(Callback.from(() -> onOpenSuccess(callback), (t) -> onOpenFail(callback, t)));
+        }
+
+        private void onOpenSuccess(Callback callback)
+        {
+            System.err.println(toString() + " onOpenSuccess()");
+
+            boolean failServer2Proxy = false;
+            Throwable failure = null;
+            synchronized (lock)
+            {
+                switch (state)
+                {
+                    case CONNECTING:
+                        state = State.OPEN;
+                        break;
+
+                    case FAILED:
+                        failure = error;
+                        failServer2Proxy = true;
+                        break;
+
+                    default:
+                        failure = new IllegalStateException();
+                        break;
+                }
+            }
+
+            if (failServer2Proxy)
+                server2Proxy.fail(failure, callback);
+            else if (failure != null)
+                callback.failed(failure);
```

Comment:
I think that if you end up in the `default` case, you really want to fail the `serverToProxy` as well.
If that's the case, the 2 failure cases become one and the code simplifies.